src/mgmt_conn.c

```diff
@@ -717,6 +717,181 @@ internal_snapshot(char *snap)
+ * If its NULL, it creates one and returns it
+ * Its the responsibility of caller to release the reference.
+ */
+struct json_object *
+uzfs_zvol_get_snap_jmap_ref(zvol_info_t *zinfo, int *err)
+{
+	boolean_t freeit = B_FALSE;
+	struct json_object *jmap;
+	int error;
+
+	*err = 0;
+	if (zinfo->snap_map == NULL) {
+		jmap = uzfs_zvol_fetch_snapshot_jmap(zinfo, &error);
+
+		if (error) {
+			*err = error;
+			return (NULL);
+		}
+
+		pthread_mutex_lock(&zinfo->snap_map_mutex);
+		if (zinfo->snap_map == NULL)
+			zinfo->snap_map = jmap;
+		else
+			freeit = B_TRUE;
+		pthread_mutex_unlock(&zinfo->snap_map_mutex);
+
+		if (freeit == B_TRUE)
+			json_object_put(jmap);
+	}
+	jmap = NULL;
+	pthread_mutex_lock(&zinfo->snap_map_mutex);
+	if (zinfo->snap_map == NULL)
+		*err = EAGAIN;
+	else
+		jmap = json_object_get(zinfo->snap_map);
+	pthread_mutex_unlock(&zinfo->snap_map_mutex);
+
+	return (jmap);
+}
+
+/*
+ * This fn adds/deletes snapname to zinfo's snap_map in json format.
+ * If errors, sets snap_map to NULL as update failed
+ * This fn is currently called only if snapshot is created / deleted through
+ * istgt, but, not through CLI
+ */
+int
+uzfs_zvol_update_snapshot_jmap(zvol_info_t *zinfo, char *snapname,
+    boolean_t add)
+{
+	int ret = 0;
+	struct json_object *jmap;
+
+	jmap = uzfs_zvol_get_snap_jmap_ref(zinfo, &ret);
+	if (ret == 0) {
+		if (add)
+			json_object_object_add(jmap, snapname, NULL);
+		else
+			json_object_object_del(jmap, snapname);
+		json_object_put(jmap);
+	} else {
+		pthread_mutex_lock(&zinfo->snap_map_mutex);
+		if (zinfo->snap_map != NULL)
+			json_object_put(zinfo->snap_map);
+		zinfo->snap_map = NULL;
+		pthread_mutex_unlock(&zinfo->snap_map_mutex);
+	}
+	return (0);
+}
+
+/*
+ * This fn gets the list of snapshots in json format, and adds it to nvl.
+ * `json_object_to_json_string_ext` is a single threaded fn, i.e., interally,
+ * this library uses a single buffer
+ * https://html.developreference.com/article/23032444/, and thus,
+ * lock has been added while using the jarray which is ref of snap_list jobj.
+ *
+ * Caller need to take refcount on zinfo so that snap_list is safe to use.
+ */
+int
+uzfs_zvol_add_nvl_snapshot_list(zvol_info_t *zinfo, nvlist_t *nvl)
+{
+	struct json_object *jmap, *jobj;
+	int error = 0;
+	const char *json_string;
+
+	jmap = uzfs_zvol_get_snap_jmap_ref(zinfo, &error);
+	if (error != 0)
+		return (error);
+
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "name",
+	    json_object_new_string(zinfo->name));
+	json_object_object_add(jobj, "snaplist", jmap);
+
+	json_string = json_object_to_json_string_ext(jobj,
+	    JSON_C_TO_STRING_PLAIN);
+	fnvlist_add_string(nvl, "listsnap", json_string);
+
+	/* This will decrement refcount of jmap as well */
+	json_object_put(jobj);
+
+	return (0);
+}
+
+/*
+ * This fn returns list of snapshots by walking through datasets in string
+ * format. Returns 0 on success.
+ * This will return EAGAIN if the replica is NOT connected to target.
```

Comment:
This will be called through mgmt connection only.

Reply:
correct..